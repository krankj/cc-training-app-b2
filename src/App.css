.container {
  padding: 3rem;
}

.tableHeader > span {
  font-weight: bold;
}

label {
  margin-right: 1rem;
}

input {
  padding: 1rem;
  font-size: 1.2rem;
}

.submitButton {
  margin-left: 1rem;
  padding: 0.5rem;
}

.submitButton:hover {
  cursor: pointer;
}

.disButton:hover {
  cursor: not-allowed;
}
