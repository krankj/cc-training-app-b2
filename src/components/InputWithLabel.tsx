const InputWithLabel = ({ onChange, value, children, id }: any) => {
  //event handler function
  const handleChange = (e: any) => {
    onChange(e);
  };

  return (
    <>
      <label htmlFor={id}>{children}</label>
      <input id={id} value={value} onChange={handleChange} />
    </>
  );
};

export default InputWithLabel;
