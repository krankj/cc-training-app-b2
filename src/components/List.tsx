import { memo } from "react";
import styles from "./List.module.css";

type Story = {
  title: string;
  url: string;
  author: string;
  points: number;
  objectID: number;
  num_comments: number;
};

type ItemProps = {
  item: Story;
  onDelete: (id: number) => void;
};

export type Stories = Array<Story>;

type ListProps = {
  stories: Stories;
  onDelete: (id: number) => void;
};

const List = memo(({ stories, onDelete }: ListProps) => {
  return (
    <div>
      <div className="tableHeader">
        <span>Title</span>
        <span>URL</span>
        <span>Author</span>
        <span>Action</span>
      </div>
      {stories.map((item: any) => (
        <Item key={item.objectID} item={item} onDelete={onDelete} />
      ))}
    </div>
  );
});

const Item = ({ item, onDelete }: ItemProps) => (
  <div>
    <span className={styles.itemTitle}>{item.title}</span>
    <span className={styles.itemUrl}>{item.url}</span>
    <span>{item.author}</span>
    <button
      onClick={() => {
        onDelete(item.objectID);
      }}
    >
      Delete
    </button>
  </div>
);

export default List;

export { Item };
