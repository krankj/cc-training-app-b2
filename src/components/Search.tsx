const Search = ({ onChange }: any) => {
  //event handler function
  const handleChange = (e: any) => {
    onChange(e);
  };

  return (
    <div>
      <label htmlFor="search">Search</label>
      <input id="search" onChange={handleChange} />
    </div>
  );
};

export default Search;
