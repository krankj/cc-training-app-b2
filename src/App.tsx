import "./App.css";
import List from "./components/List";
import InputWithLabel from "./components/InputWithLabel";
import useSemiPersistentState from "./hooks/useSemiPersistentState";

const list = [
  {
    title: "React",
    url: "https://reactjs.org/",
    author: "Jordan Walke",
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: "Learn React",
    url: "https://reactjs.org/",
    author: "Jordan Walke",
    num_comments: 3,
    points: 4,
    objectID: 1,
  },
  {
    title: "Redux",
    url: "https://redux.js.org/",
    author: "Dan Abramov, Andrew Clark",
    num_comments: 2,
    points: 5,
    objectID: 2,
  },
];

function App() {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("", "searchTerm");
  //callback handler
  const handleSearchChange = (e: any) => {
    setSearchTerm(e.target.value);
  };

  const filteredList: any = list.filter((item: any) =>
    item.title.includes(searchTerm)
  );

  return (
    <div className="container">
      <h1>Hacker Stories</h1>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onChange={handleSearchChange}
      >
        Search
      </InputWithLabel>
      <List stories={filteredList} />
    </div>
  );
}
export default App;
