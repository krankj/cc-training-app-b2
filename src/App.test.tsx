import React from "react";
import { fireEvent, render, screen, act } from "@testing-library/react";
import SearchForm from "./components/SearchForm";
import axios from "axios";
import App, { storiesReducer } from "./App";
import { Item } from "./components/List";

describe("Sample test suite", () => {
  test("expecting true to be true", () => {
    const isCold = true;
    expect(isCold).toBe(true);
  });
});

const storyOne = {
  title: "React",
  url: "https://reactjs.org/",
  author: "Jordan Walke",
  num_comments: 3,
  points: 4,
  objectID: 0,
};

const storyTwo = {
  title: "Redux",
  url: "https://redux.js.org/",
  author: "Dan Abramov, Andrew Clark",
  num_comments: 2,
  points: 5,
  objectID: 1,
};

const stories = [storyOne, storyTwo];

describe("Test stories reducer", () => {
  test("remove story from reducer", () => {
    const state = { data: stories, isLoading: false, isError: false };
    const action = { type: "REMOVE_STORY", payload: storyTwo.objectID };
    const result = storiesReducer(state, action);
    const newState = { data: [storyOne], isLoading: false, isError: false };
    expect(result).toStrictEqual(newState);
  });
});

describe("Item component", () => {
  test("render item component", () => {
    render(<Item item={storyOne} onDelete={() => {}} />);
    screen.debug();
  });

  test("assert if title exists", () => {
    render(<Item item={storyOne} onDelete={() => {}} />);
    expect(screen.getByText("React")).toBeInTheDocument();
  });

  test("button exists", () => {
    render(<Item item={storyOne} onDelete={() => {}} />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("handle delete click", () => {
    const handleDeleteItem = jest.fn();
    render(<Item item={storyOne} onDelete={handleDeleteItem} />);
    fireEvent.click(screen.getByRole("button"));
    expect(handleDeleteItem).toHaveBeenCalledTimes(1);
  });
});

describe("Search form", () => {
  test("render the search form", () => {
    const searchProps = {
      searchTerm: "React",
      onSearch: jest.fn(),
      onSubmit: jest.fn(),
    };

    render(<SearchForm {...searchProps} />);
    expect(screen.getByDisplayValue("React")).toBeInTheDocument();
    expect(screen.getByLabelText(/search/i)).toBeInTheDocument();
    fireEvent.change(screen.getByDisplayValue("React"), {
      target: { value: "Redux" },
    });
    expect(searchProps.onSearch).toHaveBeenCalledTimes(1);
    fireEvent.submit(screen.getByRole("button"));
    expect(searchProps.onSubmit).toHaveBeenCalledTimes(1);

    screen.debug();
    expect(screen.getByDisplayValue("Redux")).toBeInTheDocument();
  });

  test("snapshot testing", () => {
    const searchProps = {
      searchTerm: "React",
      onSearch: jest.fn(),
      onSubmit: jest.fn(),
    };

    const { container } = render(<SearchForm {...searchProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});

//INTEGRATION TESTING

jest.mock("axios");

describe("Integration testing of axios", () => {
  test("test loading data", async () => {
    const resolvedPromise = Promise.resolve({ data: { hits: stories } });
    axios.get.mockImplementationOnce(() => resolvedPromise);
    render(<App />);
    screen.debug();
    expect(screen.getByText(/Loading/i)).toBeInTheDocument();
    await act(() => resolvedPromise);
    screen.debug();
    expect(screen.queryByText(/Loading/i)).toBeNull();
  });

  test("fail loading data", async () => {
    const rejectedPromise = Promise.reject();
    axios.get.mockImplementationOnce(() => rejectedPromise);

    render(<App />);
    try {
      await act(() => rejectedPromise);
    } catch {
      expect(screen.getByText(/went wrong/i)).toBeInTheDocument();
    }
  });
});
