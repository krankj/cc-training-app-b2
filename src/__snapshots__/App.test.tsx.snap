// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search form snapshot testing 1`] = `
<form>
  <label
    for="search"
  >
    Search
  </label>
  <input
    id="search"
    value="React"
  />
  <button
    class="submitButton "
    type="submit"
  >
    Submit
  </button>
</form>
`;
